File: cs50/week6/pset/sentimental-mario-more/mario.py
"""
Prompts user for an int height and creates a Mario-style pyramid with
# characters.
"""


def main():
    """Take user input and print pyramid."""

    # Get height of pyramid.
    height = get_height('Height: ')

    # Print pyramid.
    print_pyramid(height)


def get_height(prompt: str) -> int:
    """Return user-input positive integer pyramid height."""

    # Continually re-prompt if height is not between 1 and 8, inclusive.
    while True:
        try:
            h = int(input(prompt))
        except ValueError:
            print('error: height must be integer')
        else:
            if 1 <= h and h <= 8:
                return h

        # Reprompt user if height is not between 1 and 8.
        print('error: height must be between 1 and 8, inclusive')


def print_pyramid(height: int):
    """Print pyramid with height input."""

    # Print input number of rows, starting with row 1 to simplify calc.
    for row in range(1, height + 1):

        # Print decreasing spaces.
        print(' ' * (height - row), end='')

        # Then print increasing # starting with 1.
        print('#' * row, end='')

        # Then print a double space.
        print('  ', end='')

        # Then print the same amount of # as before.

        print('#' * row, end='')

        # Print new line after each row
        print()


if __name__ == "__main__":
    main()

File: cs50/test/js/mario.js
// Airbnb JavaScript style guide: https://github.com/airbnb/javascript/

/**
 * Prompts user for an int height and creates a Mario-style pyramid with
 * # characters.
 */

// Import the readline-sync module to interact with user via console.
const readlineSync = require('readline-sync');

// Main function.
const main = function mainFunction () {
  // Get height of half-pyramid.
  let userHeight = getHeight('Height: ');

  // Print pyramid.
  console.log();
  printPyramid(userHeight);
}

// Return user-input positive integer pyramid height.
const getHeight = function getHeightFunction (prompt) {
  let height;

  // Continually reprompt if height is not an integer between 1 and 8.
  while (true) {
    height = Number(readlineSync.question(prompt));
    if (!Number.isInteger(height)) {
      console.log('error: height must be an integer.\n');
      continue;
    } else if (height < 1 || height > 8) {
      console.log('error: height must be between 1 and 8, inclusive.\n');
      continue;
    }

    return height;
  }

}

// Print pyramid with height input.
const printPyramid = function printPyramidFunction (height) {
  // Print input number of rows, starting with row 1 to simplify calc.
  for (let row = 1; row < height + 1; row++) {
    // Initialize printer string.
    let printer = '';

    // Add decreasing spaces.
    printer += ' '.repeat(height - row);

    // Then add increasing # starting with 1.
    printer += '#'.repeat(row);

    // Then add a double space.
    printer += '  ';

    // Then add the same amount of # as before.
    printer += '#'.repeat(row);

    // Print row.
    console.log(printer);
  }
}

main();
